### PR TITLE
Playback and Record for Azure Storage Blob

### PR DESCRIPTION
**Pull request by alzimmermsft**
Fixes part of #4713 

**Note: ~1000 of the files are for playback, those can be ignored**

- Creates a playback and record framework for Azure Storage Blobs (uses Groovy and Spock for unit testing).
- Adds playback records (will be added after draft review).
- Updates Azure Core Test to fix edge cases and extend functionality needed by storage tests.
- Adds `NetworkCallException` to handle serialization and deserialization of exceptions thrown in the pipeline during playback and record.
- Adds recording handling for pipeline exceptions.
- Default serialization of the response body no longer assumes compression, fixes exception thrown by record policy.
- Adds OffsetDateTime creation in `TestResourceNamer`.

### CODE REVIEW

**Comment by conniey** [comment]
I started to look at this, but the diff is 1024files, I don't know if I can manage that many. 😄

File: sdk/core/azure-core-test/src/main/java/com/azure/core/test/policy/RecordNetworkCallPolicy.java [hunk 1]
```diff
@@ -104,40 +126,77 @@ public RecordNetworkCallPolicy(RecordedData recordedData) {
...
             });
         }
     }
+
+    private String redactUserDelegationKey(String content) {
+        if (!content.contains("UserDelegationKey")) {
+            return content;
+        }
+
+        content = redactionReplacement(content, DELEGATIONKEY_KEY_PATTERN.matcher(content), Base64.getEncoder().encodeToString("REDACTED".getBytes(StandardCharsets.UTF_8)));
```
Comment (sima-zhu) [inline sdk/core/azure-core-test/src/main/java/com/azure/core/test/policy/RecordNetworkCallPolicy.java]: Since you have the helper method, it is better to simplify the second param to string

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/APISpec.groovy [hunk 2]
```diff
@@ -96,222 +94,345 @@ class APISpec extends Specification {
...
+    static SharedKeyCredential alternateCredential
+    static SharedKeyCredential blobCredential
+    static SharedKeyCredential premiumCredential
+    static TestMode testMode
 
-    /*
-    URLs to various kinds of accounts.
-     */
-    BlobServiceClient primaryServiceURL
+    BlobServiceClient primaryServiceClient
```
Comment (sima-zhu) [inline sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/APISpec.groovy]: I'd suggest we use the same way to define the variable. Xstore team suggests to use **def** as much as possible.

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/APISpec.groovy [hunk 3]
```diff
@@ -96,222 +94,345 @@ class APISpec extends Specification {
...
-        return ctx.getCurrentFeature().name.replace(' ', '').toLowerCase()
-    }
+        primaryServiceClient = setClient(primaryCredential)
+        alternateServiceClient = setClient(alternateCredential)
+        blobStorageServiceClient = setClient(blobCredential)
+        premiumServiceClient = setClient(premiumCredential)
 
-    def generateContainerName() {
-        generateContainerName(specificationContext, iterationNo, entityNo++)
+        cu = primaryServiceClient.getContainerClient(generateContainerName())
```
Comment (sima-zhu) [inline sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/APISpec.groovy]: cu -> containerClient

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/APISpec.groovy [hunk 4]
```diff
@@ -96,222 +94,345 @@ class APISpec extends Specification {
...
+        @Override
+        ProxyOptions get() {
+            return new ProxyOptions(ProxyOptions.Type.HTTP, new InetSocketAddress("localhost", 8888))
+        }
+    }
 
-        primaryServiceURL = getGenericServiceURL(primaryCreds)
-        cu = primaryServiceURL.getContainerClient(containerName)
-        cu.create()
+    def generateContainerName() {
```
Comment (sima-zhu) [inline sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/APISpec.groovy]: Is this still in use?

File: sdk/core/azure-core-test/src/main/java/com/azure/core/test/http/PlaybackClient.java [hunk 5]
```diff
@@ -141,7 +160,12 @@ private String applyReplacementRule(String text) {
     }
 
     private static String removeHost(String url) {
-        URI uri = URI.create(url);
-        return String.format("%s?%s", uri.getPath(), uri.getQuery());
+        UrlBuilder urlBuilder = UrlBuilder.parse(url);
+
+        if (urlBuilder.query().containsKey("sig")) {
```
[the remark this answers is not in the document]
Comment (alzimmermsft) [reply sdk/core/azure-core-test/src/main/java/com/azure/core/test/http/PlaybackClient.java]: It's useful to keep the rest of the SAS token the same as that is what is being matched in the playback client, it could help catch scenarios where we accidentally break the generation of it.

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/APISpec.groovy (hunk 2 above)
Comment (alzimmermsft) [reply sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/APISpec.groovy]: I'd prefer not to use `def` here as it cannot infer type as it is assigned during runtime and it makes it much harder to program as IntelliSense is lost.

File: sdk/core/azure-core-test/src/main/java/com/azure/core/test/models/NetworkCallError.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+     */
+    public NetworkCallError(Throwable throwable) {
+        this.throwable = throwable;
+        this.className = throwable.getClass().getName();
+    }
+
+    /**
+     * @return the thrown throwable as the class it was thrown as by converting is using its class name.
+     */
+    public Throwable get() {
```
[the remark this answers is not in the document]
Comment (alzimmermsft) [reply sdk/core/azure-core-test/src/main/java/com/azure/core/test/models/NetworkCallError.java]: Not entirely certain what you mean by this, do you want it to return `<T extends Throwable> T`?

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/APISpec.groovy (hunk 4 above)
Comment (alzimmermsft) [reply sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/APISpec.groovy]: Yes, it's used in 30 places based on IntelliJ's find all references.

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/APISpec.groovy (hunk 3 above)
Comment (alzimmermsft) [reply sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/APISpec.groovy]: I think James is changing this in another PR and I don't want to cause a ton of tiny merge conflicts.

File: sdk/core/azure-core-test/src/main/java/com/azure/core/test/policy/RecordNetworkCallPolicy.java (hunk 1 above)
Comment (alzimmermsft) [reply sdk/core/azure-core-test/src/main/java/com/azure/core/test/policy/RecordNetworkCallPolicy.java]: I'd rather handle the encoding here as the other redactions don't require an encoding.

File: sdk/core/azure-core-test/src/main/java/com/azure/core/test/models/NetworkCallError.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (sima-zhu) [reply sdk/core/azure-core-test/src/main/java/com/azure/core/test/models/NetworkCallError.java]: Talked offline. Will make the exception class generic.

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/SASTest.groovy [hunk 7]
```diff
@@ -669,9 +610,9 @@ class SASTest extends APISpec {
         e.getMessage().contains(parameter)
 
         where:
-        version | creds       || parameter
-        null    | primaryCreds | "version"
-        "v"     | null         | "sharedKeyCredentials"
+        version | creds             || parameter
```
Comment (sima-zhu) [inline sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/SASTest.groovy]: I guess the table format of `||` is typo right here. Can you make a singe `|`?

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobAPITest.groovy [hunk 8]
```diff
@@ -652,9 +651,9 @@ class BlockBlobAPITest extends APISpec {
         validateBlobProperties(response, cacheControl, contentDisposition, contentEncoding, contentLanguage, contentMD5, contentType)
 
         where:
-        cacheControl | contentDisposition | contentEncoding | contentLanguage | contentMD5                                                   | contentType
-        null         | null               | null            | null            | null                                                         | null
-        "control"    | "disposition"      | "encoding"      | "language"      | MessageDigest.getInstance("MD5").digest(defaultData.array()) | "type"
+        cacheControl | contentDisposition | contentEncoding | contentLanguage | contentMD5                                                                | contentType
```
Comment (sima-zhu) [inline sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobAPITest.groovy]: The format is weird. Remove extra space

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlobOutputStreamTest.groovy [hunk 9]
```diff
@@ -0,0 +1,74 @@
+package com.azure.storage.blob
+
+import com.azure.storage.common.Constants
+import spock.lang.Ignore
+
+class BlobOutputStreamTest extends APISpec {
+    private static int FOUR_MB = 4 * Constants.MB
+
+    @Ignore
```
Comment (sima-zhu) [inline sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlobOutputStreamTest.groovy]: When do you want them run?
Could you make them run in live or leave the reason of ignore.

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlobAPITest.groovy [hunk 10]
```diff
@@ -682,7 +660,9 @@ class BlobAPITest extends APISpec {
         setup:
         String leaseID = setupBlobLeaseCondition(bu, receivedLeaseID)
 
-        Thread.sleep(16000) // Wait for the lease to expire to ensure we are actually renewing it
+        // If running in live mode wait for the lease to expire to ensure we are actually renewing it
+        sleepIfRecord(16000)
```
Comment (sima-zhu) [inline sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlobAPITest.groovy]: It seems a long time to run. Can live test finish in 60 mins with the time span like this?
Might need to ignore in live tests

File: sdk/core/azure-core-test/src/main/java/com/azure/core/test/http/PlaybackClient.java [hunk 11]
```diff
@@ -121,7 +124,20 @@ public HttpClient port(int i) {
                 }
             }
 
-            bytes = rawBody.getBytes(StandardCharsets.UTF_8);
+            String contentType = networkCallRecord.response().get("Content-Type");
+
+            // octet-stream's are written to disk using Arrays.toString() which creates an output such as "[12, -1]".
+            if (contentType != null && contentType.contains("octet-stream")) {
```
Comment (conniey) [inline sdk/core/azure-core-test/src/main/java/com/azure/core/test/http/PlaybackClient.java]: "application/octet-stream" mime-type is well known. I'd use `.equalsIgnoreCase()` because they are also case-insensitive.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/APISpec.groovy [hunk 12]
```diff
@@ -95,222 +93,345 @@ class APISpec extends Specification {
...
+        if (testMode == TestMode.RECORD) {
+            builder.addPolicy(interceptorManager.getRecordPolicy())
         }
 
-        try {
-            blobStorageServiceURL = getGenericServiceURL(getGenericCreds("BLOB_STORAGE_"))
+        return builder.credential(credential).buildBlobClient()
+    }
+
+    BlobClient getBlobClient(SharedKeyCredential credential, String endpoint, String blobName) {
```
Comment (sima-zhu) [inline sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/APISpec.groovy]: Since you have several different clients depend on same builder. Could you make the helper method for the builder instead of client?

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/APISpec.groovy [hunk 13]
```diff
@@ -437,83 +555,47 @@ class APISpec extends Specification {
...
-    This method returns a stub of an HttpResponse. This is for when we want to test policies in isolation but don't care
-     about the status code, so we stub a response that always returns a given value for the status code. We never care
-     about the number or nature of interactions with this stub.
-     */
-
-    def getStubResponse(int code) {
-        return Stub(HttpResponse) {
-            statusCode() >> code
+    // Only sleep if test is running in live mode
+    def sleepIfRecord(long milliseconds) {
```
Comment (sima-zhu) [inline sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/APISpec.groovy]: Why you take million sec here? Can we take second?

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlobAPITest.groovy (hunk 10 above)
Comment (sima-zhu) [reply sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlobAPITest.groovy]: I thought the value is second. Ignore the comments here

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlobOutputStreamTest.groovy (hunk 9 above)
Comment (alzimmermsft) [reply sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlobOutputStreamTest.groovy]: These were always ignored, when they were in a JUnit test class the `@Test` was commented out.

Will look to enable these in a follow-up PR.

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/DownloadResponseMockFlux.java [hunk 14]
```diff
@@ -31,15 +31,16 @@
...
+
         switch (this.scenario) {
             case DR_TEST_SCENARIO_SUCCESSFUL_ONE_CHUNK:
-                this.scenarioData = APISpec.getRandomData(512 * 1024);
+                this.scenarioData = apiSpec.getRandomData(512 * 1024);
                 break;
             case DR_TEST_SCENARIO_SUCCESSFUL_MULTI_CHUNK:
             case DR_TEST_SCENARIO_SUCCESSFUL_STREAM_FAILURES:
-                this.scenarioData = APISpec.getRandomData(1024);
+                this.scenarioData = apiSpec.getRandomData(1024);
```
Comment (sima-zhu) [inline sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/DownloadResponseMockFlux.java]: I think the static method is enough. It is weird to take in the APISpec

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlobAPITest.groovy (hunk 10 above)
Comment (alzimmermsft) [reply sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlobAPITest.groovy]: `Thread.sleep` is in milliseconds, running the blob tests in live mode doesn't come close to hitting the 60 minute timeout.

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/APISpec.groovy (hunk 13 above)
Comment (alzimmermsft) [reply sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/APISpec.groovy]: That is what `Thread.sleep` takes.

File: sdk/core/azure-core-test/src/main/java/com/azure/core/test/policy/RecordNetworkCallPolicy.java [hunk 15]
```diff
@@ -102,40 +123,77 @@ public RecordNetworkCallPolicy(RecordedData recordedData) {
...
         }
 
-        String contentType = response.headerValue("content-type");
+        String contentType = response.headerValue("Content-Type");
         if (contentType == null) {
             return Mono.just(responseData);
-        } else if (contentType.contains("json") || response.headerValue("content-encoding") == null) {
-            return response.bodyAsString().switchIfEmpty(Mono.just("")).map(content -> {
-                responseData.put("Body", content);
+        } else if (contentType.contains("octet-stream")) {
```
Comment (conniey) [inline sdk/core/azure-core-test/src/main/java/com/azure/core/test/policy/RecordNetworkCallPolicy.java]: There are a lot of special string literals "octet-stream", "Body", "Content-Encoding". It would be nice to consolidate these in either a constants file or at the top of the class so we don't miss updating one.

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/DownloadResponseMockFlux.java (hunk 14 above)
Comment (alzimmermsft) [reply sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/DownloadResponseMockFlux.java]: The method can't be static as it requires test context to properly record the data.

File: sdk/core/azure-core-test/src/main/java/com/azure/core/test/policy/RecordNetworkCallPolicy.java (hunk 15 above)
Comment (conniey) [reply sdk/core/azure-core-test/src/main/java/com/azure/core/test/policy/RecordNetworkCallPolicy.java]: Same comment about application/octet-stream

File: sdk/core/azure-core-test/src/main/java/com/azure/core/test/http/PlaybackClient.java (hunk 11 above)
Comment (conniey) [reply sdk/core/azure-core-test/src/main/java/com/azure/core/test/http/PlaybackClient.java]: in the RFC, headers are case insensitive. It is by convention that you see "Content-Type" rather than "content-type". But they amount to the same thing.

Comment (sima-zhu) [reply sdk/core/azure-core-test/src/main/java/com/azure/core/test/http/PlaybackClient.java]: Playback client is not validator. Is it better to ignore the case for all the hard-code string checking?

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/SASTest.groovy (hunk 7 above)
Comment (alzimmermsft) [reply sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/SASTest.groovy]: Spock allows for `||` in the data table, I don't think it causes any differences in the test run but its mean to be a visual indicator of what the test is compared to in the end.

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlobAPITest.groovy [hunk 16]
```diff
@@ -850,9 +830,9 @@ class BlobAPITest extends APISpec {
 
         where:
         leaseTime | breakPeriod | remainingTime
-        -1        | null        | 0
-        -1        | 20          | 25
-        20        | 15          | 16
+            -1        | null        | 0
```
[the remark this answers is not in the document]
Comment (alzimmermsft) [reply sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlobAPITest.groovy]: Fixed

File: sdk/core/azure-core-test/src/main/java/com/azure/core/test/http/PlaybackClient.java (hunk 11 above)
Comment (conniey) [reply sdk/core/azure-core-test/src/main/java/com/azure/core/test/http/PlaybackClient.java]: RecordNetworkCallPolicy encounters the same issues. I added a comment about that.

Comment (alzimmermsft) [reply sdk/core/azure-core-test/src/main/java/com/azure/core/test/http/PlaybackClient.java]: While RFC headers are case-insensitive the data structure, `Map`, we use to store this information in playback isn't. This is another cleanup we should do to either use `HttpHeaders` or a case-insensitive `Map` implementation.

Comment (conniey) [reply sdk/core/azure-core-test/src/main/java/com/azure/core/test/http/PlaybackClient.java]: SO has revealed that there is a way to have case insensitive key comparison. One thing that .NET provides over HashMap.
https://stackoverflow.com/a/22336599

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/APISpec.groovy (hunk 12 above)
Comment (alzimmermsft) [reply sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/APISpec.groovy]: Will look into this change after the PR goes through.

File: sdk/core/azure-core-test/src/main/java/com/azure/core/test/http/PlaybackClient.java (hunk 11 above)
Comment (alzimmermsft) [reply sdk/core/azure-core-test/src/main/java/com/azure/core/test/http/PlaybackClient.java]: Updated to the case ignores check.

File: sdk/core/azure-core-test/src/main/java/com/azure/core/test/policy/RecordNetworkCallPolicy.java (hunk 15 above)
Comment (alzimmermsft) [reply sdk/core/azure-core-test/src/main/java/com/azure/core/test/policy/RecordNetworkCallPolicy.java]: Updated

File: sdk/core/azure-core-test/src/main/java/com/azure/core/test/policy/RecordNetworkCallPolicy.java (hunk 1 above)
Comment (sima-zhu) [reply sdk/core/azure-core-test/src/main/java/com/azure/core/test/policy/RecordNetworkCallPolicy.java]: I mean you can have the signature like
redactionReplacement(String content, String regex, String replacement)
Not a big fan of method call in signature. Hard to log and debug. The third param  makes sense to me as some key might use other way to encode.

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/DownloadResponseMockFlux.java (hunk 14 above)
Comment (sima-zhu) [reply sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/DownloadResponseMockFlux.java]: Not a big blocker for this. Might be better to add the randomData method in TestRandomNamer
